rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -1964,27 +1489,27 @@ pub fn can_push(msg: &RequestOrResponse, available_memory: i64) -> Result<(), us
     }
 }
 
-/// Returns the memory required to push `req` onto an input or output queue.
-/// This is the maximum of `MAX_RESPONSE_COUNT_BYTES` (to be reserved for a
-/// response) and `req.count_bytes()` (if larger).
+/// Returns the guaranteed response memory required to push `req` onto an input
+/// or output queue.
+///
+/// For best-effort requests, this is always zero. For guaranteed response
+/// requests, this is the maximum of `MAX_RESPONSE_COUNT_BYTES` (to be reserved
+/// for a guaranteed response) and `req.count_bytes()` (if larger).
 pub fn memory_required_to_push_request(req: &Request) -> usize {
-    req.count_bytes().max(MAX_RESPONSE_COUNT_BYTES)
-}
+    if req.deadline != NO_DEADLINE {
+        return 0;
+    }
```

Comment:
Relatedly, if we go for option (b) we also have to think about whether out of memory should be an error where we immediately reject upon induction. But probably this is not required if there is a round of shedding after induction.

Reply:
> This would make it more difficult (if not downright impossible) to generate reject signals

I suppose you mean reject messages here, right? -- for reject signals this is a non-concern as we would never pick up more than msg_begin of the remote subnet (i.e., the index up to which we will garbage collect when the header of that stream is inducted) + 10k once Christian's work is completed.

Also note that not being able to induct because of a lack of memory is just one reason for a reject upon induction. It could also be related to a stopped/non-existent canister etc. While I agree with your argumentation that just going ahead and inducting best-effort messages pretending they require 0 bytes and potentially shedding them later, the other cases are problematic, and we either need to handle them as per suggestion (a) or (b).

Reply:
> I suppose you mean reject messages here, right?

No, I actually meant reject signals. If we induct a best-effort request; immediately do a load-shedding round that results in shedding the best-effort request we just inducted; then it's already too late to generate a reject signal.

> the other cases are problematic, and we either need to handle them as per suggestion (a) or (b).

If you mean the "stopped/non-existent canister etc." cases, we already generate reject responses for those (regardless of whether they are guaranteed response or best-effort). And, when we transition to reject signals, we will automatically switch to producing a signal instead.

We could change the behavior to silently drop best-effort requests that we failed to induct instead, but seeing as how reject signals are on the horizon, there isn't much benefit in taking that approach.

Exceeding the memory pool size is different, though. If we don't explicitly check it during induction (and I really don't think we have to), then we can just induct the request (assuming the canister is there, the queue has space, etc.) and, if it so happens, shed it soon after, without producing a reject response (which is what the logic in this MR would do). At some later point, the callback will expire and the originating subnet will generate a timeout reject response.

Reply:
> No, I actually meant reject signals. [...]

Ok I guess we are on the same page that this is a non concern once we have reject signals. 

> If you mean the "stopped/non-existent canister etc." cases, we already generate reject responses for those (regardless of whether they are guaranteed response or best-effort).

Precisely. And once we have best effort messages enabled this will become a problem because there can be too many such reject responses around unless we either do (a) or (b) or possibly something else. This is the problem I'm talking about above.

> Exceeding the memory pool size is different, though. [...]

Agreed. I realized that memory size is unrelated to my concern midway in the discussion. So the only remaining issue is the one related to stopped/non-existent cansiters and possibly other cases where we'd produce a reject response.

Reply:
> there can be too many such reject responses around unless we either do (a) or (b) or possibly something else.

Got it. I had forgotten about that aspect (which is why we introduced reject signals to begin with).

I guess this means that we are now in wild agreement.

Reply:
Kind of.

There is no need to explicitly reject incoming best-effort requests if they push us over the memory limit. We can just induct them; shed load until we're back under the limit; and if we happened to drop them in the process, thne so be it. For best-effort requests that actually fail to be inducted (e.g. because they are addressed to nonexistent or stopped canisters), we could technically silently drop them (and produce an accept signal); but ideally, once we have reject signals, we should be friendly and produce a reject signal for them.

For guaranteed response requests, we must produce reject responses / reject signals for, regardless of cause. Even though there's a relatively low upper limit on how many outstanding requests a subnet can produce, a malicious subnet could send out as many as it wants to. So for them too it will be safer and cleaner when we have support for reject signals: regardless of how many (best-effort or guaranteed response) requests the other subnet enqueues in its stream for us, we will only pick up, say, 10k (so if we already have signals for 3k messages, we will pull at most 7k more). Worst-case scenario, we end up with 10k reject signals in our outgoing stream.